File: python/tracing/app/receiver.py
import sys
sys.path.append('./')
from flask import Flask, request
import json
import time
from lib.tracing import OtelTracer

tracer = OtelTracer(service_name='python-tracing-demo-receiver', service_version='0.0.0')
tracer.initialize()

app = Flask(__name__)

@app.route("/job", methods=['POST'])
def job():
    data = request.get_json()

    # Deserialize context from b3
    parent_context = tracer.from_b3(data['b3'])

    # Create child span from extracted context
    # child_span = tracer.tracer.start_span("job", parent_context=parent_context)
    child_span = tracer.span(name="job", parent_context=parent_context)

    time.sleep(1)

    child_span.end()
    return json.dumps({
        'status': 'ok'
    })


if __name__ == '__main__':
    app.run(host='127.0.0.1', threaded=True)

File: nodejs/tracing/app/receiver.ts
import express from 'express';
import { OtelTracer } from '../lib/otel-tracer';


const tracer = new OtelTracer('nodejs-tracing-demo-receiver', '0.0.0');
tracer.initialize();

const app = express();
app.use(express.json());


app.post('/job', async (req, res) => {
    console.log(`Received job with b3 ${JSON.stringify(req.body)}`);

    // Parse the parent context from the b3 string passed in the request body.
    //
    // NOTE: normally b3 strings are sent in HTTP headers. We're using the body here to show
    // that this technique is not HTTP specific. Any transport mechanism can be used: GRPC, MQTT, Sockets.
    const parentContext = tracer.fromB3(req.body.b3);
    const span = tracer.span('job', {}, parentContext);

    // Simulate work
    await new Promise((resolve) => setTimeout(resolve, Math.random() * 1000));

    span.end();
    res.status(200).json({
        status: 'OK',
    })
})


app.listen(3000, () => {
    console.log('Listening on http://localhost:3000')
})